src/System.IO.FileSystem/src/System/IO/FindEnumerableFactory.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.IO
+{
+    internal static class FindEnumerableFactory
+    {
+        internal static void NormalizeInputs(ref string directory, ref string expression)
+        {
+            if (PathHelpers.IsPathRooted(expression))
+                throw new ArgumentException(SR.Arg_Path2IsRooted, nameof(expression));
+
+            // We always allowed breaking the passed in directory and filter to be separated
+            // any way the user wanted. Looking for "C:\foo\*.cs" could be passed as "C:\" and
+            // "foo\*.cs" or "C:\foo" and "*.cs", for example. As such we need to combine and
+            // split the inputs if the expression contains a directory separator.
+            //
+            // We also allowed for expression to be "foo\" which would translate to "foo\*".
+
+            ReadOnlySpan<char> directoryName = PathHelpers.GetDirectoryNameNoChecks(expression.AsReadOnlySpan());
+
+            if (directoryName.Length != 0)
+            {
+                // Need to fix up the input paths
+                directory = PathHelpers.CombineNoChecks(directory, directoryName);
+                expression = expression.Substring(directoryName.Length + 1);
+            }
+
+            // Historically we always treated "." as "*"
+            if (string.IsNullOrEmpty(expression)
+                || expression == "." || expression == "*.*")
+                expression = "*";
+        }
+
+        internal static FindEnumerable<string> UserFiles(string directory,
+            string expression = "*",
+            bool recursive = false)
+        {
+            expression = DosMatcher.TranslateExpression(expression);
+
+            bool predicate(ref RawFindData findData)
+            {
+                return FindPredicates.NotDotOrDotDot(ref findData)
+                    && !FindPredicates.IsDirectory(ref findData)
+                    && DosMatcher.MatchPattern(expression, findData.FileName, ignoreCase: true);
+            }
+
+            return new FindEnumerable<string>(directory, recursive, FindTransforms.AsUserFullPath, predicate);
```

Comment:
Passing `FindTransforms.AsUserFullPath` and `predicate` like this is going to result in two delegate allocations each time `UserFiles` is called. Same for the methods below. It'd be nice if these delegate instances were lazily allocated and cached/reused.

You could attempt to address this by writing it as:

```c#
    	FindTransform<string> transform = (ref RawFindData findData) =>
        {
            ReadOnlySpan<char> subdirectory = findData.Directory.AsReadOnlySpan().Slice(findData.OriginalDirectory.Length);
            return PathHelpers.CombineNoChecks(findData.OriginalUserDirectory, subdirectory, findData.FileName);
        };
        
        FindPredicate predicate = (ref RawFindData findData) =>
        {
            return FindPredicates.NotDotOrDotDot(ref findData)
                && !FindPredicates.IsDirectory(ref findData)
                && DosMatcher.MatchPattern(expression, findData.FileName, ignoreCase: true);
        };
    
    	return new FindEnumerable<string>(directory, recursive, transform, predicate);
```

The C# compiler will lazily allocate and cache/reuse the transform delegate, but predicate is problematic as written above as it captures `expression` resulting in a closure allocation.

To avoid that, you'd have to change the `FindPredicate` signature to also take a `string expression` parameter, and then pass the expression to the `FindEnumerable` constructor so it can pass it along whenever it calls `predicate`, e.g.:

```c#
internal delegate bool FindPredicate(string expression, ref RawFindData findData);
```

```c#
        FindPredicate predicate = (string expr, ref RawFindData findData) =>
        {
            return FindPredicates.NotDotOrDotDot(ref findData)
                && !FindPredicates.IsDirectory(ref findData)
                && DosMatcher.MatchPattern(expr, findData.FileName, ignoreCase: true);
        };

        return new FindEnumerable<string>(directory, expression, recursive, transform, predicate);
```

```c#
        _predicate(_expression, ref findData);
```

Reply:
I can do the first part for sure. Redefining the predicate delegate is problematic I think. Not all predicates need the extra parameter and future ones will need different arguments (think all files modified after some DateTime).

Reply:
> Not all predicates need the extra parameter and future ones will need different arguments

Can `FindPredicate` take a generic `TState` that can be used to pass any extra info needed?

Alternatively, instead of using delegates for the predicate and transform you could have abstract/virtual Transform/Predicate methods on the enumerable class, and have sealed subclasses for each type of enumeration.